types/wlr_cursor.c

```diff
@@ -233,69 +233,81 @@ static struct wlr_box *get_mapping(struct wlr_cursor *cur,
 }
 
 bool wlr_cursor_warp(struct wlr_cursor *cur, struct wlr_input_device *dev,
-		double x, double y) {
+		double lx, double ly) {
 	assert(cur->state->layout);
-	bool result = false;
 
+	bool result = false;
 	struct wlr_box *mapping = get_mapping(cur, dev);
+	if (mapping) {
+		result = wlr_box_contains_point(mapping, lx, ly);
+	} else {
+		result = wlr_output_layout_contains_point(cur->state->layout, NULL,
+			lx, ly);
+	}
+
+	if (result) {
+		cursor_warp_unchecked(cur, lx, ly);
+	}
+
+	return result;
+}
 
+static void cursor_warp_closest(struct wlr_cursor *cur,
+		struct wlr_input_device *dev, double lx, double ly) {
+	struct wlr_box *mapping = get_mapping(cur, dev);
 	if (mapping) {
-		if (wlr_box_contains_point(mapping, x, y)) {
-			cursor_warp_unchecked(cur, x, y);
-			result = true;
+		if (!wlr_box_contains_point(mapping, lx, ly)) {
+			double closest_x, closest_y;
+			wlr_box_closest_point(mapping, lx, ly, &closest_x, &closest_y);
+			lx = closest_x;
+			ly = closest_y;
+		}
+	} else {
+		if (!wlr_output_layout_contains_point(cur->state->layout, NULL,
+				lx, ly)) {
+			double closest_x, closest_y;
+			wlr_output_layout_closest_point(cur->state->layout, NULL, lx, ly,
+				&closest_x, &closest_y);
+			lx = closest_x;
+			ly = closest_y;
 		}
-	} else if (wlr_output_layout_contains_point(cur->state->layout, NULL,
-				x, y)) {
-		cursor_warp_unchecked(cur, x, y);
-		result = true;
 	}
 
-	return result;
+	cursor_warp_unchecked(cur, lx, ly);
 }
 
-void wlr_cursor_warp_absolute(struct wlr_cursor *cur,
-		struct wlr_input_device *dev, double x, double y) {
+void wlr_cursor_absolute_to_layout_coords(struct wlr_cursor *cur,
+		struct wlr_input_device *dev, double x, double y,
+		double *lx, double *ly) {
 	assert(cur->state->layout);
 
 	struct wlr_box *mapping = get_mapping(cur, dev);
 	if (!mapping) {
 		mapping = wlr_output_layout_get_box(cur->state->layout, NULL);
 	}
 
-	x = x >= 0 ? mapping->width * x + mapping->x : cur->x;
-	y = y >= 0 ? mapping->height * y + mapping->y : cur->y;
+	*lx = !isnan(x) ? mapping->width * x + mapping->x : cur->x;
+	*ly = !isnan(y) ? mapping->height * y + mapping->y : cur->y;
+}
 
-	cursor_warp_unchecked(cur, x, y);
+void wlr_cursor_warp_absolute(struct wlr_cursor *cur,
+		struct wlr_input_device *dev, double x, double y) {
+	assert(cur->state->layout);
+
+	double lx, ly;
+	wlr_cursor_absolute_to_layout_coords(cur, dev, x, y, &lx, &ly);
+
+	cursor_warp_closest(cur, dev, lx, ly);
 }
 
 void wlr_cursor_move(struct wlr_cursor *cur, struct wlr_input_device *dev,
 		double delta_x, double delta_y) {
 	assert(cur->state->layout);
 
-	double x = cur->x + delta_x;
-	double y = cur->y + delta_y;
-
-	struct wlr_box *mapping = get_mapping(cur, dev);
+	double lx = !isnan(delta_x) ? cur->x + delta_x : cur->x;
```

Comment:
The negation here is odd. IMO it would be better for readability to not negate and swap the clauses (same for y)

Reply:
Hmmm, not sure about this change. I like having the important case first. But I don't like ternary operations like this in the first place, so…

Reply:
I just thought of it as a bit surprising, but if you want to keep it this way, fine by me